include/ftl/thread_local.h

```diff
@@ -0,0 +1,195 @@
+
+	ThreadLocal<T>& m_parent;
+	T& m_value;
+};
+
+
+	/**
+ * Fiber compatible version of the thread_local keyword. Acts like a pointer. Think of this as you would the thread_local keyword,
+ * not something you put in containers, but a specification of a variable itself. Non-copyable and non-movable.
+ * For proper semantics, all variables that are ThreadLocal<> must be static.
+ *
+ * @tparam T    Object type to store.
+ */
+template<class T>
+class ThreadLocal {
+private:
+	template<class VP_T>
+	struct alignas(CACHE_LINE_SIZE) ValuePadder {
+		ValuePadder() : m_value() {}
+		VP_T m_value;
+		bool m_inited = true;
+	};
+
+public:
+	/**
+	 * Default construct all T's inside the thread local variable.
+	 *
+	 * @param ts    The task scheduler to be thread local to.
+	 */
+
+#ifdef FTL_OS_WINDOWS
+	#pragma warning(push)
+	#pragma warning(disable: 4316) // I know this won't be allocated to the right alignment, this is okay.
+#endif
+	explicit ThreadLocal(TaskScheduler* ts)
+		: m_scheduler{ts},
+		  m_initalizer{},
+		  m_data{new ValuePadder<T>[ts->GetThreadCount()]} {}
+#ifdef FTL_OS_WINDOWS
+	#pragma warning(pop)
+#endif
+
+	/**
+	 * Construct all T's by calling a void factory function the first time you use your data.
+	 *
+	 * @tparam F      Type of the factory function
+	 * @param ts      The task scheduler to be thread local to
+	 * @param args    Factory function to initialize the values with.
+	 */
+	template<class F>
+	ThreadLocal(TaskScheduler* ts, F&& factory)
+	  : m_scheduler{ts},
+	    m_initalizer{std::forward<F>(factory)},
+	    m_data(static_cast<ValuePadder<T>*>(::operator new[](sizeof(ValuePadder<T>) * ts->GetThreadCount()))) {
+        for (std::size_t i = 0; i < ts->GetThreadCount(); ++i) {
+            m_data[i].m_inited = false;
+        }
+    }
+
+	ThreadLocal(ThreadLocal const& other) = delete;
+	ThreadLocal(ThreadLocal&& other) noexcept = delete;
+	ThreadLocal& operator=(ThreadLocal const& other) = delete;
+	ThreadLocal& operator=(ThreadLocal&& other) noexcept = delete;
+
+	~ThreadLocal() {
+		delete[] m_data;
+	}
+
+	/**
+	 * Get a handle to your thread's T. You must be careful that you don't jump to another thread in the mean time, but this will allow
+	 * you to prevent calls to ftl::TaskScheduler::GetCurrentThreadIndex() every single time you want to modify the object (similar to TLS optimization).
+	 * This is the recommended way of doing that over getting a reference as it allows instrumentation and debugging within the library.
+	 *
+	 * @return    Handle to the thread's version of T.
+	 */
+	ThreadLocalHandle<T> getHandle() {
+		return ThreadLocalHandle<T>{*this, **this};
+	}
+
+	T& operator*() {
+		std::size_t idx = m_scheduler->GetCurrentThreadIndex();
+		init_value(idx);
+		return m_data[idx].m_value;
+	}
+	T* operator->() {
+		std::size_t idx = m_scheduler->GetCurrentThreadIndex();
+		init_value(idx);
+		return &m_data[idx].m_value;
+	}
+	std::vector<T> getAllValues() {
```

Comment:
In the doxygen comment, can you add a note that this returns *copies* of the current values. Just to make sure the user is aware.